src/main/java/org/commcare/formplayer/application/FormController.java

```diff
@@ -200,14 +201,20 @@ public SubmitResponseBean submitForm(@RequestBody SubmitRequestBean submitReques
             try {
                 restoreFactory.setAutoCommit(false);
 
-                SimpleTimer purgeCasesTimer = FormRecordProcessorHelper.processXML(
-                        new FormplayerTransactionParserFactory(
-                                restoreFactory.getSqlSandbox(),
-                                storageFactory.getPropertyManager().isBulkPerformanceEnabled()
-                        ),
-                        formEntrySession.submitGetXml(),
-                        storageFactory.getPropertyManager().isAutoPurgeEnabled()
-                ).getPurgeCasesTimer();
+                SimpleTimer purgeCasesTimer = null;
+
+                try {
+                    purgeCasesTimer = FormRecordProcessorHelper.processXML(
+                            new FormplayerTransactionParserFactory(
+                                    restoreFactory.getSqlSandbox(),
+                                    storageFactory.getPropertyManager().isBulkPerformanceEnabled()
+                            ),
+                            formEntrySession.submitGetXml(),
+                            storageFactory.getPropertyManager().isAutoPurgeEnabled()
+                    ).getPurgeCasesTimer();
+                } catch (InvalidCaseGraphException e) {
+                    log.error("Case purge failed during form submission due to " + e.getMessage());
```

Comment:
Does the purge occur *after* the form is processed in this case? I feel like if that's the case, we should avoid submitting the form to HQ, and error out of the process as if the form failed validation after rolling back the changes.

This would prevent formplayer from submitting up a form which likely creates a cycle, preventing the need to be able to deal with it asynchronously on HQ's side (It can't catch every case, but might catch most of them).

Reply:
@ctsims That does seem like a good suggestion.The purge does happen after we have processed the form. Though I am a bit confused how would we rollback the changes caused by form processing locally once the form processing is done ?

Reply:
Since autocommit is disabled, I believe you can execute the following to rollback changes to the sqlite db

```
restoreFactory.getSQLiteDB().getConnection().rollback();
```

that call should really get abstracted *into* the restoreFactory, though.

Reply:
We could set the status on the response that indicates the form was invalid because it creates a cycle and early exit. And then that would be displayed to the user here

Reply:
Ah, Cal is correct, it looks like as long as this method exits before committing, it'll automatically roll back the transacitons

https://github.com/dimagi/formplayer/pull/948/files/fd4b7d5ab86c073b950e53bdc80eea60b73716f7#diff-d44694a41dfd4c703a35f74e71c28499ca3bb16b86ad10e9353c6737ae39cf91R260